pkg/controllers/nodeoverlay/validation/controller.go

```diff
@@ -0,0 +1,248 @@
+			// overlay that is applied
+			if conflictingPriceOverlay == "" && conflictingCapacityOverlay == "" {
+				// if both are not defined, price will be set to nil
+				price := lo.Ternary(overlay.Spec.Price == nil, overlay.Spec.PriceAdjustment, overlay.Spec.Price)
+				offeringMap := map[string]*string{}
+				if price != nil {
+					for _, of := range offerings {
+						offeringMap[of.Requirements.String()] = price
+					}
+				}
+				store[it.Name] = InstanceTypeUpdate{
+					overlayName: overlay.Name,
+					price:       price,
+					offering:    offeringMap,
+					capacity:    overlay.Spec.Capacity,
+					weight:      overlay.Spec.Weight,
+				}
+			} else {
+				overlaysWithConflict = append(overlaysWithConflict, overlay.Name, conflictingPriceOverlay, conflictingCapacityOverlay)
+			}
+		}
+	}
+
+	return overlaysWithConflict
+}
+
+// isOfferingsForCompatibleInstanceType will validate that an instance type matches a set of node overlay requirements
+// if true, the set of Compatible offering
+func isOfferingsForCompatibleInstanceType(it *cloudprovider.InstanceType, overlayReqA scheduling.Requirements) cloudprovider.Offerings {
+	if it.Requirements.Compatible(overlayReqA) == nil {
+		if overlayReqA.Keys().HasAny(v1.WellKnownLabelsForOfferings.UnsortedList()...) {
+			matchingOfferings := cloudprovider.Offerings{}
+			for _, of := range it.Offerings {
+				if of.Requirements.Compatible(overlayReqA) == nil {
+					matchingOfferings = append(matchingOfferings, of)
+				}
+			}
+			return matchingOfferings
+		} else {
+			return it.Offerings
+		}
+	}
+	return nil
+}
+
+func isPriceAlreadyUpdated(store map[string]InstanceTypeUpdate, it *cloudprovider.InstanceType, offerings cloudprovider.Offerings, overlay v1alpha1.NodeOverlay) string {
+	overlayPriceChange := lo.Ternary(overlay.Spec.Price == nil, overlay.Spec.PriceAdjustment, overlay.Spec.Price)
+	updateIt, foundUpdatedInstanceType := store[it.Name]
+	if !foundUpdatedInstanceType {
+		return ""
+	}
+
+	for _, of := range offerings {
+		offeringPrice, foundOffering := updateIt.offering[of.Requirements.String()]
+		if lo.FromPtr(overlay.Spec.Weight) == lo.FromPtr(updateIt.weight) {
+			if (overlayPriceChange != nil && lo.FromPtr(offeringPrice) == lo.FromPtr(overlayPriceChange)) || !foundOffering {
+				updateIt.price = overlayPriceChange
+				updateIt.offering[of.Requirements.String()] = updateIt.price
+			} else {
+				return updateIt.overlayName
+			}
+		}
+	}
+
+	return ""
+}
+
+func isCapacityAlreadyUpdated(store map[string]InstanceTypeUpdate, it *cloudprovider.InstanceType, overlay v1alpha1.NodeOverlay) string {
+	updateIt, foundUpdatedInstanceType := store[it.Name]
+
+	if foundUpdatedInstanceType && lo.FromPtr(overlay.Spec.Weight) == lo.FromPtr(updateIt.weight) && overlay.Spec.Capacity != nil && updateIt.capacity != nil {
+		resource := findConflictingResources(overlay.Spec.Capacity, updateIt.capacity)
+		if len(resource) == 0 {
+			updateIt.capacity = overlay.Spec.Capacity.DeepCopy()
```

Comment:
We should be merging the capacity in the `InstanceTypeUpdate` with the overlay spec not overriding it, right?

Reply:
Yeah, updated this section to be a merge semantic and added a test for it